demo/read_write_demo.py

```diff
@@ -0,0 +1,24 @@
+"""Example of reading and writing data using dynamicio."""
+from pathlib import Path
+
+from pandera import Field, Float, SchemaModel, String
+from pandera.typing import Series
+
+from dynamicio.handlers.file import ParquetFileResource
+
+
+### Example 2 ###
+class BarSchema(SchemaModel):
+    column_a: Series[String] = Field(unique=True)
+    column_b: Series[String] = Field(nullable=False)
+    column_c: Series[Float] = Field(gt=1000)
+    # column_d: Series[Float] = Field(lt=1000)
+
+    class Config:
+        strict = "filter"
+
+
+TEST_RESOURCES = Path(__file__).parent / "data"
+resource = ParquetFileResource(path=TEST_RESOURCES / "input/bar.parquet").read(pa_schema=BarSchema)
+df = resource.read()
```

Comment:
Any plans for lazy loading? Or should I just call `.read()` at the moment of usage?

Reply:
I would be quite happy to add lazy loading as a cached property. Last time I tried to do this I was told "that's not the way" though.

Reply:
As the decorator? `@cached_property`? Repeated reads will not perform any IO, yes, but what I'm referring to is the following situation:

Usually we preload all the data at the beginning of the task (which creates a huge memory footprint). Then, at some point, we use the actual dataset, and then we don't use it for the rest of the task. The problem is memory footprint. The solution is twofold:

1. Lazy load the dataset. Only fetch it when the dataset is being used
2. Then perform garbage collection to free up memory

Any plans to bake the above into `v5`?

Reply:
Could you give an example? 

I'm not sure how to do what you're asking apart from calling the `@cached_property` / `.read()` at the point where the data is needed and not before that. That means we can either pass down the resource definition or just import and use it in relevant places. Or are you suggesting some dataframe object that only gets loaded once any actions are performed on it? That would be cool, not sure how to technically solve that atm.

Reply:
>Or are you suggesting some dataframe object that only gets loaded once any actions are performed on it

Yep, that's what I mean by lazy loading

Reply:
That would be very cool, but definitely an extra :D

Reply:
Now you've got me thinking xD One hacky way to do it might be to wrap every single function that df would have with some loading function? Or maybe if any function is called from this, then replace itself with the loaded df. But it would have to include all dunder methods pretty much. haxhaxhax... hmm

Reply:
@Christakou came up with a hack for this through variable scoping/garbage collection. It may very well be possible to bake this in cleanly 🤷

Reply:
Would be an amazingly cool feature, you could make ay LazyDataFrame object that subclasses a pandas dataframe, but that implements this lazy loading mechanics by hecking if the object has been _hydrated before a function call, and triggers the IO at that point